pkg/networkservice/acl/acl_parser.go

```diff
@@ -0,0 +1,273 @@
+	if !ok {
+		return errors.New("invalid action")
+	}
+
+	rule.IsPermit = a
+	return nil
+}
+
+func addPrefixes(parsed map[string]string, rule *acl_types.ACLRule) {
+	dst, dstOk := parsed[dstPrefix]
+	if !dstOk {
+		rule.DstPrefix = ip_types.Prefix{}
+	} else {
+		dstpref, err := ip_types.ParsePrefix(dst)
+		if err != nil {
+			rule.DstPrefix = ip_types.Prefix{}
+		} else {
+			rule.DstPrefix = dstpref
+		}
+	}
+
+	src, srcOk := parsed[srcPrefix]
+	if !srcOk {
+		rule.SrcPrefix = ip_types.Prefix{}
+	} else {
+		srcpref, err := ip_types.ParsePrefix(src)
+		if err != nil {
+			rule.SrcPrefix = ip_types.Prefix{}
+		} else {
+			rule.SrcPrefix = srcpref
+		}
+	}
+}
+
+var protocolMap = map[string]ip_types.IPProto{
+	"hopopt": ip_types.IP_API_PROTO_HOPOPT,
+	"icmp":   ip_types.IP_API_PROTO_ICMP,
+	"igmp":   ip_types.IP_API_PROTO_IGMP,
+	"tcp":    ip_types.IP_API_PROTO_TCP,
+	"udp":    ip_types.IP_API_PROTO_UDP,
+	"gre":    ip_types.IP_API_PROTO_GRE,
+	"esp":    ip_types.IP_API_PROTO_ESP,
+	"ah":     ip_types.IP_API_PROTO_AH,
+	"icmp6":  ip_types.IP_API_PROTO_ICMP6,
+	"eigrp":  ip_types.IP_API_PROTO_EIGRP,
+	"ospf":   ip_types.IP_API_PROTO_OSPF,
+	"sctp":   ip_types.IP_API_PROTO_SCTP,
+}
+
+func addProtocol(parsed map[string]string, rule *acl_types.ACLRule) error {
+	proto, ok := parsed[protocol]
+	if !ok {
+		return errors.New("protocol is missing")
+	}
+
+	p, ok := protocolMap[proto]
+	if !ok {
+		return errors.New("invalid protocol")
+	}
+
+	rule.Proto = p
+	return nil
+}
+
+func addSrcPortOrIcmpType(parsed map[string]string, rule *acl_types.ACLRule) error {
+	if rule.Proto == ip_types.IP_API_PROTO_ICMP || rule.Proto == ip_types.IP_API_PROTO_ICMP6 {
+		first, last, err := findNumberPairByKeys(parsed, icmpTypeFirst, icmpTypeLast)
+		if err != nil {
+			return err
+		}
+
+		rule.SrcportOrIcmptypeFirst = first
+		rule.SrcportOrIcmptypeLast = last
+
+		return nil
+	}
+
+	first, last, err := findNumberPairByKeys(parsed, srcPortFirst, srcPortLast)
+	if err != nil {
+		return err
+	}
+
+	rule.SrcportOrIcmptypeFirst = first
+	rule.SrcportOrIcmptypeLast = last
+
+	return nil
+}
+
+func addDstPortOrIcmpCode(parsed map[string]string, rule *acl_types.ACLRule) error {
+	if rule.Proto == ip_types.IP_API_PROTO_ICMP || rule.Proto == ip_types.IP_API_PROTO_ICMP6 {
+		first, last, err := findNumberPairByKeys(parsed, icmpCodeFirst, icmpCodeLast)
+		if err != nil {
+			return err
+		}
+
+		rule.DstportOrIcmpcodeFirst = first
+		rule.DstportOrIcmpcodeLast = last
+
+		return nil
+	}
+
+	first, last, err := findNumberPairByKeys(parsed, dstPortFirst, dstPortLast)
+	if err != nil {
+		return err
+	}
+
+	rule.DstportOrIcmpcodeFirst = first
+	rule.DstportOrIcmpcodeLast = last
+
+	return nil
+}
+
+func addTCPFlags(parsed map[string]string, rule *acl_types.ACLRule) {
+	tcpFlagsMask, ok := parsed[tcpFlagsMask]
+	if !ok {
+		return
+	}
+	tcpFlagsVal, ok := parsed[tcpFlagsValue]
+	if !ok {
+		return
+	}
+
+	mask, err := strconv.Atoi(tcpFlagsMask)
+	if err != nil {
```

Comment:
Probably it would be better to return an error in such case?

Reply:
done